Handle restarting

Currently, every time upgrade new version, need to manually restart the container.

```
sftp | java.io.IOException: Cannot run program "/usr/sbin/service": error=2, No such file or directory
sftp |     at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1142)
sftp |     at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
sftp |     at java.base/java.lang.Runtime.exec(Runtime.java:590)
sftp |     at java.base/java.lang.Runtime.exec(Runtime.java:449)
sftp |     at crushftp.server.ServerStatus.restart_crushftp(ServerStatus.java:4750)
sftp |     at crushftp.server.ServerStatus$14.run(ServerStatus.java:4694)
sftp |     at com.crushftp.client.Worker2$1.run(Worker2.java:72)
sftp |     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
sftp |     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
sftp |     at java.base/java.lang.Thread.run(Thread.java:831)
sftp | Caused by: java.io.IOException: error=2, No such file or directory
sftp |     at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
sftp |     at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:313)
sftp |     at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
sftp |     at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1109)
sftp |     ... 9 more
sftp | 2021-09-28T02:14:21+0000 Stopping...
sftp | 2021-09-28T02:14:21+0000 Cannot find PID
```

--- a/crushftp/Dockerfile
+++ b/crushftp/Dockerfile
@@ -9,6 +9,7 @@ FROM $SOURCE AS src
 FROM base as build
 ARG STARTUP=startup.sh
 ADD $STARTUP /src/var/opt/startup.sh
+ADD service.sh /src/var/opt/service.sh
 WORKDIR /src/tmp
 COPY --from=src src/crushftp.zip .
 ARG CRUSHFTP_VERSION=10.1.0_62
@@ -25,6 +26,7 @@ RUN apk upgrade --no-cache \
 FROM deps as final
 COPY --from=build /src /
 WORKDIR /var/opt
+RUN ln -s /var/opt/service.sh /usr/sbin/service
 ENTRYPOINT [ "/bin/bash", "startup.sh" ]
 CMD ["-c"]
 

--- a/crushftp/service.sh
+++ b/crushftp/service.sh
@@ -1,0 +1,46 @@
+#!/usr/bin/env bash
+
+CRUSH_DIR="/var/opt/crushftp"
+SERVICE_LOG=$CRUSH_DIR/service.log
+
+log() {
+    echo "$(date '+%FT%T%z') $@" >> $SERVICE_LOG
+}
+
+log_n() {
+    echo -n "$(date '+%FT%T%z') $@" >> $SERVICE_LOG
+}
+
+COMMAND_ARGS="$@"
+log "Service execute: '$COMMAND_ARGS'"
+
+get_pid() {
+    unset CRUSH_PID
+    CRUSH_PID="`ps -a | grep "java" | grep "$CRUSH_DIR/CrushFTP.jar" | awk '{print $1}'`"
+}
+
+if [[ "$COMMAND_ARGS" == "crushftp restart" ]] ; then
+    log "[Restart] Stopping..."
+    get_pid
+    #echo "CRUSH_PID:  $CRUSH_PID"
+    if [[ -z "$CRUSH_PID" ]] ; then    
+        log "Cannot find PID"
+        exit 1
+    fi
+
+    log_n "[Restart] Shutting down CrushFTP... "
+    kill $CRUSH_PID
+    ret_val=$?
+    if [ ${ret_val} -ne 0 ] ; then
+        echo FAIL
+        log "Could not kill PID"
+        exit 1
+    fi
+    echo OK >> $SERVICE_LOG
+
+    log_n "[Restart] Starting CrushFTP... "
+    java -Xmx512m -jar $CRUSH_DIR/CrushFTP.jar -d & >/dev/null 2>&1
+    echo OK >> $SERVICE_LOG
+
+    exit 0
+fi

--- a/crushftp/startup.sh
+++ b/crushftp/startup.sh
@@ -72,18 +72,22 @@ echo "# Password:   $ADMIN_PASSWORD"
 echo "########################################"
 
 get_pid() {
+    unset CRUSH_PID
     CRUSH_PID="`ps -a | grep "java" | grep "$CRUSH_DIR/CrushFTP.jar" | awk '{print $1}'`"
 }
+
+unset SIGTERM_HANDLING
 
 #get_pid
 #log PID $CRUSH_PID has been started!
 
 # SIGTERM-handler
 term_handler() {
+    SIGTERM_HANDLING=1
     log "Stopping..."
     get_pid
     #echo "CRUSH_PID:  $CRUSH_PID"
-    if [[ -z "$CRUSH_PID" ]]; then    
+    if [[ -z "$CRUSH_PID" ]] ; then    
         log Cannot find PID
         exit 1
     fi
@@ -91,7 +95,7 @@ term_handler() {
     log_n "Shutting down CrushFTP... "
     kill $CRUSH_PID
     ret_val=$?
-    if [ ${ret_val} -ne 0 ]; then
+    if [ ${ret_val} -ne 0 ] ; then
         echo FAIL
         log "Could not kill PID"
         exit 1
@@ -106,6 +110,18 @@ trap 'term_handler' SIGTERM
 
 while true
 do
-    sleep 60m &
+    sleep 30s &
     wait $!
+    
+    if [[ -z "$SIGTERM_HANDLING" ]] ; then
+        #log "idle..."
+
+        get_pid        
+        if [[ -z "$CRUSH_PID" ]] ; then
+            log "Cannot find PID, try to start CrushFTP again"
+            log_n "Try starting CrushFTP... "
+            java -Xmx512m -jar $CRUSH_DIR/CrushFTP.jar -d & >/dev/null 2>&1
+            echo OK
+        fi
+    fi
 done

--- a/crushftp/startup.sh
+++ b/crushftp/startup.sh
@@ -21,6 +21,12 @@ log_n() {
     echo -n "$(date '+%FT%T%z') $@"
 }
 
+launch_CrushFTP() {
+    log_n "Starting CrushFTP... "
+    java -Xmx512m -jar $CRUSH_DIR/CrushFTP.jar -d & >/dev/null 2>&1
+    echo OK
+}
+
 log "Initializing..."
 
 if [[ ! -d $CRUSH_DIR ]] ; then
@@ -54,9 +60,7 @@ else
     ADMIN_PASSWORD="****************"
 fi
 
-log_n "Starting CrushFTP... "
-java -Xmx512m -jar $CRUSH_DIR/CrushFTP.jar -d & >/dev/null 2>&1
-echo OK
+launch_CrushFTP
 
 log "Waiting server starting..."
 
@@ -119,9 +123,7 @@ do
         get_pid        
         if [[ -z "$CRUSH_PID" ]] ; then
             log "Cannot find PID, try to start CrushFTP again"
-            log_n "Try starting CrushFTP... "
-            java -Xmx512m -jar $CRUSH_DIR/CrushFTP.jar -d & >/dev/null 2>&1
-            echo OK
+            launch_CrushFTP
         fi
     fi
 done

